### PR TITLE
Support floating prerelease and stable packages at the same time with the same version range (enable absolute latest scenario) (*-*)

### LINKED ISSUE #912: Support pre-release packages with floating versions
Consider a project.json specification such as

```
{
   "A: "1.0.*"
}
```

If the only packages in the feed are prerelease packages such as `A.1.0.0-alpha`, `A.1.0.1-beta`, the restore fails stating it was unable to resolve package `A`.

------------
[anangaur] Adding the requirement as evolved during the discussions:

As Noel, who uses NuGet packages in PackageReference based projects,

- I would like to use floating versions that only resolves to stable versions of the package (Current behavior)
- I would like to use floating version that can resolve to the latest version even if the latest happens to be a pre-release version.
- I should be able to use both the braces/regex formats i.e. 1.* or [1.0.0, 2.0.0) that can include pre-release versions
- I would like to additionally define whether I want to resolve to rc, beta, alpha or all pre-releases.

------------
**Link to spec** - https://github.com/NuGet/Home/wiki/Support-pre-release-packages-with-floating-versions
**Status** - reviewed

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/912
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: Spec at https://github.com/NuGet/Home/blob/dev/designs/FloatingStableAndPrerelease.md
You can test out the implementation at the following [link](https://nugettoolsdev.azurewebsites.net/5.5.0-floating.7611/find-best-version-match?versionRange=5.*-rc.*&versions=4.9.0%0D%0A5.1.0%0D%0A5.2.0-beta.1%0D%0A5.2.0-rc.3%0D%0A6.0.0-rc.1)

## Testing/Validation

Tests Added: Yes  
Reason for not adding tests:  
Validation:  Lots of unit tests, covering all scenarios.

### CODE REVIEW

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs
Comment (awesley) [inline src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Lines 123/124, do you mean to do `charArray[1]` and `charArray[2]`?  Along with other cases where you're testing against `*-*`, you might compare it against a well-named field/const to improve readability. 

Additionally, we do not test this case.

File: src/NuGet.Core/NuGet.Versioning/FloatRange.cs
Comment (awesley) [inline src/NuGet.Core/NuGet.Versioning/FloatRange.cs:119]: Lines 119 and 126 you use `FloatBehavior` instead of `_floatBehavior`, which is otherwise used.  You may consider choosing one or the other so readers do not wonder if there's a difference (which as far as I can tell it is inconsequential).

File: src/NuGet.Core/NuGet.Versioning/FloatRange.cs
Comment (awesley) [inline src/NuGet.Core/NuGet.Versioning/FloatRange.cs]: Might add a comment that this is just testing for `*-*`.

**Review by awesley — COMMENTED** [review]
🍕

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Nice catch, it parses to the correct thing anyways, so I think I'll just remove it. 
No point in special casing it. It's a minor perf benefit, but it's arguably harder to maintain.
There's a test for it, but I now realize the test name is...off. 
https://github.com/NuGet/NuGet.Client/pull/3247/files#diff-4512545cf1140ab61c395477ce2352a2R400

File: src/NuGet.Core/NuGet.Versioning/FloatRange.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Versioning/FloatRange.cs:119]: You are correct, it's inconsequential. 
I have switched everything to an auto property for consistency.

File: src/NuGet.Core/NuGet.Versioning/FloatRange.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Versioning/FloatRange.cs]: I think it's self descriptive...thoughts @NuGet/nuget-client

**Comment by nkolev92** [comment]
Thanks for the review @awesley. :)

File: src/NuGet.Core/NuGet.Versioning/FloatRange.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Versioning/FloatRange.cs]: How about `else if (versionString.Equals("*-*", StringComparison.Ordinal))`? I'd be surprised if the 4 boolean operators `&&`'d together have a noticable performance improvement. If the performance is important, this check should be done before calling `versionString.IndexOf` and `versionString.LastIndexOf`.

Similarly, the first condition above could be `if (versionString.Equals("*", StringComparison.Ordinal))`.

~~If you do both of these, you could mode `firstStarPosition` and `lastStarPosition` into the `else` block regardless of performance concerns.~~ there is no else block. The else block is a lie.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (cristinamanum) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Can range verification added in the test as well?

Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Can you clarify what specifically you are referring to?

File: src/NuGet.Core/NuGet.Versioning/FloatRange.cs
Comment (cristinamanum) [inline src/NuGet.Core/NuGet.Versioning/FloatRange.cs:194]: Is "&& versionString.IndexOf('+') == -1" needed? How *-*+001 will be resolved?

**Review by cristinamanum — APPROVED** [review]

File: src/NuGet.Core/NuGet.Versioning/FloatRange.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Versioning/FloatRange.cs:194]: `+` is only valid in a version's metadata, which not part of the version identity.
As such we do not allow it in floating versions,

File: src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Pretty sure you intended the array indexes to be 0,1,2, respectively.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Ahh, my bad, I thought I removed that. 

As this is currently, it's dead code, and there are tests covering this scenario, so removing the special case is safe, so I'll just do that.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Apparently I left it in the FloatRange, so I'll just do the same.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Versioning/VersionRangeFactory.cs]: Removed. It was the actually the same feedback as https://github.com/NuGet/NuGet.Client/pull/3247/files#r383501926, so not sure how this happened. 
Thanks for catching it.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Per coding guidelines, we technically shouldn't use `var` here. If you agree, there's a bunch of places in your tests you'd need to change 😔

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Can we move them now?

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs:366]: Named _Parse_ but is testing `TryParse`. I'd like to see both tested to verify that Parse returns a non-null `range`.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Named _Parse_ but is testing `TryParse`. I'd like to see both tested to verify that Parse returns a null `range`.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeFloatParsingTests.cs
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeFloatParsingTests.cs:198]: What about an Empty string? Perhaps we should test that whitespace, random character, integer, etc are considered invalid rather than an unhandled exception?

**Review by donnie-msft — CHANGES_REQUESTED** [review]

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: I think it's obvious what the return type here is.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: I'll remove that comment.

They are probably fine in either place. There are tests for testing similar functionality in both classes.
Given how many tests there are > 600 lines, I'm inclined to just leave them here.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Added an extra assertion and duplicated the same test cases for both Parse and TryParse

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs:366]: Same as below: Added an extra assertion and duplicated the same test cases for both Parse and TryParse

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeFloatParsingTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeFloatParsingTests.cs:198]: Empty and white space strings were not handled by the code, but that was always the case. 
I have added an extra change to handle that more gracefully. 

The reality is that floating range parse is never invoked in production code and the VersionRange.Parse method has a try catch for this.

**Comment by nkolev92** [comment]
@donnie-msft Addressed feedback in the latest commit. 
Resolved the accepted comments, replied to the other 2.

**Review by donnie-msft — CHANGES_REQUESTED** [review]
> @donnie-msft Addressed feedback in the latest commit.
> Resolved the accepted comments, replied to the other 2.

Thanks. Looks good, but I'll let someone else approve to overrule on the coding guideline.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: The return type is a VersionRange. 
I can fix up the test name cause it seems those might be causing the confusion. 
I started off by copying test templates and seems like I've just made it worse for this particular class.

**Comment by donnie-msft** [comment]
> @donnie-msft Addressed feedback in the latest commit.
> Resolved the accepted comments, replied to the other 2.

Thanks. Looks good, but I'll let someone else approve to overrule on the coding guideline.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (donnie-msft) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: (edit: moved to this thread from Review comment)

I disagree. Nothing indicates that a FloatRange is being returned. Pretty confident @dtivel would point to our coding guideline # 10.

`The following are incorrect:

var flavor = fruit.GetFlavor();`

Comment (donnie-msft) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: It's not particularly that it's confusing, but it's a misuse of `var` per our guidelines.

Comment (donnie-msft) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Find and replace "var  range" should be pretty easy at least 😀

Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: > I disagree. Nothing indicates that a FloatRange is being returned. 

It's not. The return type is a VersionRange.

Will shift the tests around and improve the test names, but I stand by that the below is imo obvious and does not violate the rule. 

```cs
    var range = VersionRange.Parse("1.1.1.*-*");
```

Others can weigh in.  @NuGet/nuget-client

**Comment by nkolev92** [comment]
I have change the test names and moved them around in better classes as per what I suggested in the TODO that @donnie-msft commented on https://github.com/NuGet/NuGet.Client/pull/3247#discussion_r391334092.

PTAL.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (donnie-msft) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Ok, my Find of the Parse method lead me to the wrong place. I'm good with `Parse` being a `var` now that I see it's returning the same type. 👍

**Review by donnie-msft — APPROVED** [review]

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Sounds good. I've created a PR to clarify it https://github.com/NuGet/NuGet.Client/pull/3294. 

Please indicate whether you'll do further review by providing further feedback/approving/dismissing the review as stale.

Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Looks like you already did that. The update did not get pushed in my GH view, lol. 
A refresh did it.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (dominoFire) [inline test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: Looks like an extra line.

File: src/NuGet.Core/NuGet.Versioning/VersionRange.cs
Comment (dominoFire) [inline src/NuGet.Core/NuGet.Versioning/VersionRange.cs:212]: Is it better to store `_floatRange?.FloatBehavior` in a temp variable and make all comparisons from there?

File: src/NuGet.Core/NuGet.Versioning/FloatRange.cs
Comment (dominoFire) [inline src/NuGet.Core/NuGet.Versioning/FloatRange.cs]: nit: Can we count the dots without creating the array?

**Review by dominoFire — COMMENTED** [review]
Looks good. Just style and variable assigments comments.

File: src/NuGet.Core/NuGet.Versioning/FloatRange.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Versioning/FloatRange.cs]: Yeah, think looping over the string as an array is the fastest. 
Not sure how measurable it is, but I can change it.

File: src/NuGet.Core/NuGet.Versioning/VersionRange.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Versioning/VersionRange.cs:212]: It just changes readability, there's not functional change in that. 
Personally I'd prefer to just keep it as is in order to avoid unnecessarily cluttering the history.

File: test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs]: removed

**Comment by nkolev92** [comment]
I broke my tests somehow :(

**Comment by nkolev92** [comment]
I had an off by 2 error. 

Will merge once tests are green.

**Review by dominoFire — APPROVED** [review]

**Pull request merged**
